### PR TITLE
PREAPPS-5867 : Adding HAB APIs.

### PR DESCRIPTION
**Pull request by BhavinBathani**
We are going to allow user to choose contacts from HAB with new contact-chooser dialog.
HAB related API calls needs to be added for ModernUI.
GetHABRequest API is implmented.

### CODE REVIEW

**Review by auminpatel — APPROVED** [review]

File: src/schema/schema.graphql [hunk 1]
```diff
@@ -3102,6 +3149,14 @@ type Query {
 		traverseMountpoints: Boolean
 		folder: GetFolderFolderInput
 	): Folder
+	getHAB(
+		habRootGroupId: ID
+	): HabGroup
+	getDistributionListMembers(
+		limit: Int,
```
Comment (itsmeprashant) [inline src/schema/schema.graphql:3156]: Do you want it to change to Float? 
This question is in the context of the issue we faced recently with our customer.

**Review by itsmeprashant — APPROVED** [review]

File: src/schema/schema.graphql (hunk 1 above)
Comment (silentsakky) [reply src/schema/schema.graphql:3156]: I doubt we will pass limit which is very high

File: src/schema/schema.graphql [hunk 2]
```diff
@@ -1125,6 +1125,43 @@ type Folder {
...
+
+type DlGroupMembers {
+	groupMember: [DlGroupMember]
+}
+
+type HabGroup {
+	name: String
+	id: String
+	parentHabGroupId: ID
+	seniorityIndex: Int
```
Comment (silentsakky) [inline src/schema/schema.graphql:1147]: can this contain value which is larger then 2lac, if yes then convert it to Float

File: src/batch-client/index.ts [hunk 3]
```diff
@@ -967,6 +967,32 @@ export class ZimbraBatchClient {
...
 
+	public getHAB = (habRootGroupId: string) =>
+		this.jsonRequest({
+			name: 'GetHAB',
+			body: {
+				habRootGroupId
+			},
+			namespace: Namespace.Account
+		}).then(res => {
+			return get(res, 'ou.0') || [];
```
Comment (silentsakky) [inline src/batch-client/index.ts:978]: implicit return

File: src/schema/schema.graphql [hunk 4]
```diff
@@ -1125,6 +1125,43 @@ type Folder {
...
 
+type DlAttrs {
+	commonName: String
+	displayName: String
+	lastName: String
+}
+
+type DlGroupMember {
+	name: String
+	_attrs: DlAttrs
```
Comment (silentsakky) [inline src/schema/schema.graphql:1136]: normalize to attributes

File: src/schema/schema.graphql [hunk 5]
```diff
@@ -1125,6 +1125,43 @@ type Folder {
...
+type DlGroupMembers {
+	groupMember: [DlGroupMember]
+}
+
+type HabGroup {
+	name: String
+	id: String
+	parentHabGroupId: ID
+	seniorityIndex: Int
+	_attrs: HabGroupAttrs
```
Comment (silentsakky) [inline src/schema/schema.graphql:1148]: same as above

File: src/schema/schema.graphql (hunk 1 above)
Comment (BhavinBathani) [reply src/schema/schema.graphql:3156]: Limit would never be that high, so keeping it to Int only. I did use Int by following Zimbra API Docs.

File: src/schema/schema.graphql (hunk 2 above)
Comment (BhavinBathani) [reply src/schema/schema.graphql:1147]: seniorityIndex would never be that larger, so keeping it to Int only. I did use Int by following Zimbra API Docs.

File: src/batch-client/index.ts (hunk 3 above)
Comment (BhavinBathani) [reply src/batch-client/index.ts:978]: Fixed.

File: src/schema/schema.graphql (hunk 5 above)
Comment (BhavinBathani) [reply src/schema/schema.graphql]: Fixed.

File: src/schema/schema.graphql (hunk 4 above)
Comment (BhavinBathani) [reply src/schema/schema.graphql]: Fixed.

**Review by madhavdhuppe — APPROVED** [review]

File: src/batch-client/index.ts [hunk 6]
```diff
@@ -192,6 +194,21 @@ const hasUnreadDescendent = (folder: any): any => {
 	return false;
 };
 
+const updateGroupName = (habGroup: any) => {
+	return {
```
Comment (silentsakky) [inline src/batch-client/index.ts:198]: implicit return

File: src/batch-client/index.ts [hunk 7]
```diff
@@ -192,6 +194,21 @@ const hasUnreadDescendent = (folder: any): any => {
...
+		...habGroup,
+		name: get(habGroup, 'attributes.displayName')
+	};
+}
+
+const updateGroupNameRecur = (habGroups: any) => {
+	return habGroups.map((habGroup: any) => {
```
Comment (silentsakky) [inline src/batch-client/index.ts:205]: implicit return

**Review by lavegupta — APPROVED** [review]

File: src/batch-client/index.ts (hunk 7 above)
Comment (BhavinBathani) [reply src/batch-client/index.ts]: Fixed.

File: src/batch-client/index.ts (hunk 6 above)
Comment (BhavinBathani) [reply src/batch-client/index.ts]: Fixed.

File: src/normalize/entities.ts [hunk 8]
```diff
@@ -360,6 +360,48 @@ Folder.addMapping({
...
+
+HabGroup.addMapping({
+	_attrs: ['attributes', HabGroupAttrs],
+	habGroup: [ 'habGroups', HabGroup ]
+});
+
+export { HabGroup };
+
+export const DlAttrs = new Entity({
+	commonName: 'commonName',
```
Comment (silentsakky) [inline src/normalize/entities.ts:391]: entities which are same should not be put here

Comment (BhavinBathani) [reply src/normalize/entities.ts:391]: Fixed.

File: src/batch-client/index.ts [hunk 9]
```diff
@@ -153,6 +155,7 @@ import {
 
 import { CASTING_PREFS } from './constants';
 import { Notifier } from './notifier';
+import { some } from 'lodash';
```
Comment (silentsakky) [inline src/batch-client/index.ts:158]:
```suggestion
import some from 'lodash/some';
```

Comment (silentsakky) [reply src/batch-client/index.ts:158]: when you use new method from lodash in api-client, you will have to modify rollup.config.js to consider it as external dependency

Comment (BhavinBathani) [reply src/batch-client/index.ts]: Removed as it is unused. may be added automatically.

File: src/batch-client/index.ts [hunk 10]
```diff
@@ -969,6 +984,34 @@ export class ZimbraBatchClient {
...
 
+	public getHAB = (habRootGroupId: string) =>
+		this.jsonRequest({
+			name: 'GetHAB',
+			body: {
+				habRootGroupId
+			},
+			namespace: Namespace.Account
+		}).then(res => {
+			const habGroups = get(res, 'ou.0');
```
Comment (silentsakky) [inline src/batch-client/index.ts:995]: I don't see usage of normalize function here, don't we need it?

**Review by silentsakky — APPROVED** [review]

**Review by sachinpatel88 — APPROVED** [review]

**Pull request merged**